### ISSUE TITLE
Add syntax for create a secondary index with synchronous view updates

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#16454]
docs/cql/cql-extensions.md explains that secondary indexes can be made to have synchronous updates (the indexed can be used as soon as a write to the table finishes), but there is currently no direct syntax to do it, so you have to resort to doing `ALTER MATERIALIZED VIEW ... WITH synchronous_updates = true`  to the view backing the index, after creating the index normally. This trick works, but it is ugly.

The goal of this issue is to request adding an option directly to `CREATE INDEX` which will create an index with synchronous updates enabled.

Scylla's documentation https://opensource.docs.scylladb.com/stable/cql/secondary-indexes.html and also Cassandra's claim that `CREATE INDEX` can take a `WITH OPTIONS = <map literal>`. However, surprisingly, this is **not** what our parser allows - it allows instead `WITH <properties>`, just like CREATE TABLE does. This is probably a bug that was never intended (we don't support any property), although to be honest I prefer Scylla's (incorrect?) approach. We'll need to decide if we want to "fix" the parser to use WITH OPTIONS, or leave the nicer (but non-standard) WITH.

We could then add a `synchronous_updates = true` option to these properties or map.

Another problem is that we'd also like to have the ability to **change** this option on an already existing index.
Basically, we need an `ALTER INDEX` statement - which neither ScyllaDB nore Cassandra have now. If we add such a command, it will be very limited for now - nothing can be altered in the index except this synchronous_updates option.

**Comment by nyh** [comment]
In https://github.com/scylladb/scylladb/pull/18467#discussion_r1585259278, we noticed (CC @piodul) another reason why we need `CREATE INDEX` to be able to set materialized-view options (including `synchronous_updates`) and `DESC TABLE` must show these options for the index:
Without this, `DESC TABLE` cannot be used to perfectly recreate a table: A user might have created a table, create an index in it, and then ALTER the materialized view backing it to be (for example) synchronous. The information of this altering will not be saved by DESC TABLE.

**Comment by VAveryanov8** [comment]
I think we hit this (or very similar issue) in scylla-manager as part of the 1-1-restore effort. High level logic of restoring data once schema is applied looks like this:

1. Ensure the tomstone gc mode of tables is set to 'repair'. 
2. Drop views 
3. Download backup data to `/upload` directory on scylla nodes and call load sstables API (`nodetool refresh` )
4. Re-create views

The first step is needed (in comparison to a regular restore) because we need to prevent possible data resurrection that can happen if the backup is relatively old and contains tombstones, then after we restore data tombstones age can be older than `gc_grace_period`.

About the issue: 

Considering that [views are normal tables under the hood and they also should be repaired (as discussed [here](https://forum.scylladb.com/t/can-we-use-tombstone-gc-mode-repair-for-gsi-tables/1273) ), we need to apply step 1 to them as well. 
While we can alter tombstone gc mode of an `Index` using `ALTER MATERIALIZED VIEW ${index_name}_index ...`, we can't re-create `Index` with needed tomstone gc mode (Step 4), because as far as I understand, `CREATE INDEX ... WITH tombtone_gc = {'mode': 'repair'}` is not supported at the moment.

I think of options that we have:
1. We can try to alter the index tombstone_gc mode after we re-create it, but at that time the base table is already restored, so data from it can be flowing into the MV table of an index and if compaction kicks in during this period, then we can end up with data resurrection. **TBD**: Not sure if it's possible to execute `ALTER MATERIALIZED VIEW` on an Index that is currently being built.

So the questions are following: 
1. Is my understanding correct, and do we need to worry about possible data resurrection in views (indexes)?
2. Does it make sense to try to change tombstone gc mode of an Index right after it is re-created or it can be too late in this case? 
3. Could disabling compaction help somehow? 

Thanks in advance!

**Comment by karol-kokoszka** [comment]
@piodul Could you please help with having a better understanding of https://github.com/scylladb/scylladb/issues/16454#issuecomment-2865730296 ? We need it for the restore in manager.

**Comment by Michal-Leszczynski** [comment]
Even if this workaround works, we should prioritize this issue, so that we are not in danger that some future release would break this workaround. If this workaround does not work, then we should prioritize it even more.

Otherwise it could make it impossible to run restore where the user can start accessing the cluster with all of the data before the repair of the whole cluster has finished (which can take a lot of time for multi-dc vnode repair).

@piodul do you think that it's reasonable candidate for `2025.3`?

**Comment by denesb** [comment]
> > ```
> > 3. Could disabling compaction help somehow?
> > ```
> 
> I'm not sure. The view updates need to be sent via network to the view replicas, written to memtables, flushed - and only then they appear in an sstable. It is possible that tombstones might be expired in some of the intermediate stages, before they even land in the sstables. cc: [@denesb](https://github.com/denesb) do you know whether this is the case?
> 
> If they are not then I suppose this could be a workaround - disable compaction, create the index, enable tombstone_gc, reenable compaction.

Compaction is the only way to permanently garbage-collect tombstones, disabling compaction can prevent garbage collection of tombstones.
But note that garbage collectible tombstones are excluded from read repair and optionally from repair (if `enable_tombstone_gc_for_streaming_and_repair` is enabled). The cache can also garbage collect tombstones.

If your goal is to allow for a repair after restore to ensure all tombstones are propagated, then you need to disable compaction, ensure `enable_tombstone_gc_for_streaming_and_repair=false`, then do a full repair. This should ensure tombstones are properly replicated.
This workaround is not guaranteed to keep working in the future, e.g. we might add tombstone gc to memtable flush at one point, which will completely defeat this workaround. The only real solution is to alter the table's tombstone_gc settings.

**Comment by karol-kokoszka** [comment]
@mykaul This issue is connected to the fast restore (1-1 restore).

In short:
- for 1-1 restore we don't want to call the repair (saving time) and we expect it to not be needed as we just bring the exactly same state as is saved to snapshot
- we are SAFE in terms of data resurrection in the tables, because we set the `tombstone_gc_mode` to `repair`, what means we don't loose delete markers 
- but there is a problem with indexes, because Scylla doesn't give a possibility for creating the index with chosen `tombstone_gc_mode`. => it may lead to data resurrection if we don't repair the base table before recreating the view.

------------

@denesb 
> Compaction is the only way to permanently garbage-collect tombstones, disabling compaction can prevent garbage collection of tombstones.

But can we even disable the compaction on chosen table at all ? I'm aware about disabling the auto-compaction only.

--------------

So, for 1-1 restore we either.
- document the limitation regarding views or include the repair into the restore (what is against being FAST).
- fix the issue in Scylla and give possibility to set "tombstone_gc_mode = 'repair'" for views.

IMHO we should fix it in core instead of looking for work-around.

-----------------

@piodul Is the fix in core something what could be done in reasonable amount of time ? 

@mykaul can we prioritize it ?

**Comment by VAveryanov8** [comment]
> But can we even disable the compaction on chosen table at all ? I'm aware about disabling the auto-compaction only.

Just to clarify, for the workaround we need to ensure that compaction is disabled when index is created, not after.
I found that you can set `WITH compaction = {"class": "NullCompactionStrategy"}` option during table creation, but I think it's not possible for Index, because syntax `CREATE INDEX ... WITH options` is not supported :)

**Comment by mykaul** [comment]
> [@mykaul](https://github.com/mykaul) This issue is connected to the fast restore (1-1 restore).
> 
> In short:
> 
> * for 1-1 restore we don't want to call the repair (saving time) and we expect it to not be needed as we just bring the exactly same state as is saved to snapshot
> * we are SAFE in terms of data resurrection in the tables, because we set the `tombstone_gc_mode` to `repair`, what means we don't loose delete markers
> * but there is a problem with indexes, because Scylla doesn't give a possibility for creating the index with chosen `tombstone_gc_mode`. => it may lead to data resurrection if we don't repair the base table before recreating the view.
> 
> [@denesb](https://github.com/denesb)
> 
> > Compaction is the only way to permanently garbage-collect tombstones, disabling compaction can prevent garbage collection of tombstones.
> 
> But can we even disable the compaction on chosen table at all ? I'm aware about disabling the auto-compaction only.
> 
> So, for 1-1 restore we either.
> 
> * document the limitation regarding views or include the repair into the restore (what is against being FAST).
> * fix the issue in Scylla and give possibility to set "tombstone_gc_mode = 'repair'" for views.
> 
> IMHO we should fix it in core instead of looking for work-around.
> 
> [@piodul](https://github.com/piodul) Is the fix in core something what could be done in reasonable amount of time ?
> 
> [@mykaul](https://github.com/mykaul) can we prioritize it ?

If you have materialized views, do call repair, until the ideas mentioned at https://github.com/scylladb/scylladb/issues/16454#issuecomment-2919208573 are implemented. I think it's a reasonable compromise.

**Comment by denesb** [comment]
> [@denesb](https://github.com/denesb)
> 
> > Compaction is the only way to permanently garbage-collect tombstones, disabling compaction can prevent garbage collection of tombstones.
> 
> But can we even disable the compaction on chosen table at all ? I'm aware about disabling the auto-compaction only.

You can disable auto-compaction. This doesn't stop any user-requested compaction from happening, e.g. a `nodetool compact` will still trigger a major compaction.

**Comment by mykaul** [comment]
> > [@denesb](https://github.com/denesb)
> > > Compaction is the only way to permanently garbage-collect tombstones, disabling compaction can prevent garbage collection of tombstones.
> > 
> > 
> > But can we even disable the compaction on chosen table at all ? I'm aware about disabling the auto-compaction only.
> 
> You can disable auto-compaction. This doesn't stop any user-requested compaction from happening, e.g. a `nodetool compact` will still trigger a major compaction.

Reasonable. I don't expect anyone to manually execute it during restore.

**Comment by karol-kokoszka** [comment]
> > > [@denesb](https://github.com/denesb)
> > > > Compaction is the only way to permanently garbage-collect tombstones, disabling compaction can prevent garbage collection of tombstones.
> > > 
> > > 
> > > But can we even disable the compaction on chosen table at all ? I'm aware about disabling the auto-compaction only.
> > 
> > 
> > You can disable auto-compaction. This doesn't stop any user-requested compaction from happening, e.g. a `nodetool compact` will still trigger a major compaction.
> 
> Reasonable. I don't expect anyone to manually execute it during restore.

If disabling auto-compaction means "if you want to compact data, make an explicit request", then it sounds we are safe to some extent.

The scenario will be as follow:
- base table is recreated with tombstone_gc_mode = "repair"
- base table is restore but not repaired
- auto-compaction on base table is disabled
- view is being create (auto-compaction is ON by default) - it starts building 
                         --- I think there is still a risk that compaction hits here (?), although very low one
- view is altered to disable auto-compaction
- view is altered to use tombstone_gc_mode = "repair"
- auto-compaction enabled on base table and view

Assuming core will introduce the possibility of setting  the mode "repair" on view creation, I think we can implement this work-around.

**Comment by Michal-Leszczynski** [comment]
> view is being create (auto-compaction is ON by default) - it starts building
--- I think there is still a risk that compaction hits here (?), although very low one
view is altered to disable auto-compaction
view is altered to use tombstone_gc_mode = "repair"

So the problem is that we can't set tgc mode to repair on index creation. If we were able to do that, we are good to go with 1-1 restore without repair.
The things about disabling autocompaction were proposed as a workaround, but unfortunately they have the very same problem (not being able to disable autocompaction on index creation).
If we feel lucky and want to disable autocompaction right after index creation, we could just set the tgc mode to repair right after index creation and not bother with the compaction at all.
Or am I missing something?

**Comment by karol-kokoszka** [comment]
> If we feel lucky and want to disable compaction right after index creation, we could just set the tgc mode to repair right after index creation and not bother with the compaction at all.

Good point. Disabling auto-compaction is very similar to "repair" mode in out context.
It comes with the same risk.

**Comment by Michal-Leszczynski** [comment]
On the other hand, there is an [compaction manager](https://github.com/scylladb/scylladb/blob/master/api/api-doc/compaction_manager.json) API to check for running compactions and compaction history. We could try to set tgc mode to repair right after index creation and validate that there were no tombstone deleted by the compactions in the meantime. 

But to be honest, the time required for validation that it's indeed safe and implementation might just not be worth it, especially that we would get rid of this workaround with the next Scylla release. So running the repair just on index base tables before recreating indexes might be the way to go for now.

**Comment by dawmd** [comment]
A brief update on this issue so people know where we stand. I did some research on the current implementation of indexes, how they're created, and what is affected. The existing code for creating an index is scattered across different modules and pretty weird (to me at least), and I don't understand why it was done that way—it was written a long time ago and, unfortunately, the commit messages were mostly vague. Probably the most relevant piece of code is this function:

https://github.com/scylladb/scylladb/blob/6c65f72031e2571b495d3c02ae1b90cf05234e4a/db/schema_tables.cc#L1833-L1887

We have access to a pretty narrow subset of information there if we want to extend the capabilities of the statement, at least without making the code convoluted.

I prepared a few ideas how to tackle this issue and discussed them with @piodul today. We both agreed that doing a refactor and moving the logic responsible for creating the underlying view to `cql3/statements/create_index_statement.cc` is the best course of action. We have all relevant information available there and we should be able to create a cleaner and easier to understand implementation.

The main reason why I'm writing this is I'd like to make sure there are no unexpected consequences of this migration. As I said, the code was written a long time (= 8/9 years) ago, but maybe someone remembers the rationale behind the current implementation. CC @nyh, I think you're the person most familiar with the feature.

**Comment by piodul** [comment]
There is a PR in review, but we won't have enough capacity to get it properly discussed and merged this quarter. Pushing out

**Comment by nyh** [comment]
I see you are planning to "Fixes" this issue without implementing ALTER INDEX. Which I think is fine, but I went ahead and created https://github.com/scylladb/scylladb/issues/27053 about the remaining issue of ALTER INDEX.

**Comment by dawmd** [comment]
> I see you are planning to "Fixes" this issue without implementing ALTER INDEX. Which I think is fine, but I went ahead and created [#27053](https://github.com/scylladb/scylladb/issues/27053) about the remaining issue of ALTER INDEX.

Thank you, Nadav! I didn't notice that.

**Comment by bhalevy** [comment]
Why is this issue labeled with `backup and restore`?
Doesn't seem to be related, at least to B&R development...

**Comment by mykaul** [comment]
> Why is this issue labeled with `backup and restore`? Doesn't seem to be related, at least to B&R development...

See https://github.com/scylladb/scylladb/issues/16454#issuecomment-2929509647 and other discussions here why it's related.
